### PR TITLE
[dynamo] Stop writing a source onto the interned ConstantVariable(None)

### PR DESCRIPTION
**Pull request by bobrenjc93**
First of six fixes this stack needs before torch.compiler.precompile at the
top of it can capture anything.

Background, because the bug is only obvious once two facts sit next to each
other. Dynamo never handles real values while tracing; every value is wrapped
in a VariableTracker, and `ConstantVariable` is the wrapper for the ones it can
treat as compile-time constants. `ConstantVariable.create(None)` therefore does
not make a None -- it asks for the tracker that DESCRIBES None. And because
every None is indistinguishable, `create` is a cache: with no kwargs it returns
`CONSTANT_VARIABLE_NONE`, one object allocated at import and handed to every
caller in the process. The second fact is that a tracker's `source` records how
to rebuild the value, in bytecode, from the frame currently being compiled --
`LOAD_FAST x; LOAD_ATTR grad`. It is meaningful only inside that frame.

Reading a tensor attribute labels the result with an AttrSource so downstream
code can reconstruct it. `getattro_impl` did that by assignment, which is safe
only when the result is a tracker it just built. For `.grad` and
`requires_grad` it is not: the value is the shared singleton, so the assignment
stamps a name from THIS frame onto the object every later compile in the
process will also be handed. The next unrelated function to touch a None gets a
tracker claiming it can be rebuilt as `x.grad`, tries it, and dies in
`create_load` with "<name> missing" -- in code with no connection to the
culprit. The cache is fine; mutating a cached object is not, for the same
reason you do not mutate a default argument.

Clone before labelling, but only for constants. A blanket clone breaks `.data`:
the tracker it hands back is freshly created, and `VariableTracker.__init__`
rejects a source on one of those ("source must be None for
ValueMutationNew/AttributeMutationNew"), so only the in-place write gets
through. The guard above the branch already restricts constants to
grad/requires_grad, so the clone arm is exactly the singleton case.

Why the top of the stack needs it: a precompile training capture pins
`trace_autograd_ops` and seeds a zero `.grad` on every autograd target before
re-capturing, which is exactly the shape that poisons the singleton. With this
hunk reverted, 32 tests in test/test_precompile.py fail.

Test Plan:

The regression test turns on `trace_autograd_ops`, compiles a step that nulls
`.grad` before calling `backward()`, then compiles an unrelated function. The
second compile raises without the fix.

```
python test/dynamo/test_repros.py
python test/dynamo/test_misc.py
```

Authored with an AI assistant.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -4575,6 +4575,33 @@ def fn(x):
...
+
+            def step(self, x, t):
+                for p in self.model.parameters():
+                    p.grad = None
+                torch.nn.functional.mse_loss(self.model(x), t).backward()
+
+        x, t = torch.randn(5, 4), torch.randn(5, 3)
+        with torch._dynamo.config.patch(trace_autograd_ops=True):
+            step = Holder(torch.nn.Linear(4, 3)).step
+            torch.compile(step, backend="eager", fullgraph=True)(x, t)
```
Comment (aorenste) [inline test/dynamo/test_repros.py:4597]: [cleanup] My agent says:
> Pin the invariant here, and undo the poisoning so a regression doesn't take the rest of the file with it:
>
> ```python
>             torch.compile(step, backend="eager", fullgraph=True)(x, t)
>
>         none_vt = torch._dynamo.variables.ConstantVariable.create(None)
>         self.addCleanup(setattr, none_vt, "source", None)
>         self.assertIsNone(none_vt.source)
> ```
>
> `torch._dynamo.reset()` doesn't clear the module-level singleton, so a stamped source survives into every later test in the process. Running the full file with the `tensor.py` hunk reverted and the test kept: 14 unrelated tests flip pass -> fail as collateral (`test_hf_gelu_inline`, `test_primtorch`, `test_nanmean_out`, `test_weakref_callback`, `test_return_weakref`, `test_unbind_copy_out`, ...), and only one of the 15 failures names the actual problem. With the hunk applied, all pass. Asserting the singleton directly makes the real failure self-describing and keeps it contained.

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -769,7 +769,19 @@ def tp_getattro_impl(
...
+            if result.is_python_constant():
+                # A constant one is not ours to mutate: ConstantVariable(None)
+                # is a process-wide interned singleton, and method_attr_grad
+                # hands back that very object as the value side_effects holds
+                # for a pending `p.grad = None`. Writing a source onto it left
+                # every LATER compile in the process reconstructing a local from
+                # THIS frame and dying in create_load. Only grad/requires_grad
+                # reach here as constants; the rest keep the in-place write,
+                # which is what lets .data through -- its tracker is freshly
+                # created, and the constructor refuses a source on one of those.
```
Comment (aorenste) [inline torch/_dynamo/variables/tensor.py]: [cleanup] My agent says:
> Low priority. Ten lines of narrative for a three-line change, most of it a retelling of the failure that the commit message already carries. The load-bearing facts are the shared singleton and why the `else` branch can't clone:
>
> ```python
>                 # ConstantVariable.create(None) is a process-wide singleton, and
>                 # method_attr_grad hands it back for a pending `p.grad = None`; a
>                 # source written onto it leaks into every later compile. Non-constants
>                 # keep the in-place write: .data's tracker is AttributeMutationNew,
>                 # which __init__ rejects a source for.
> ```

File: torch/_dynamo/variables/tensor.py [hunk 3]
```diff
@@ -769,7 +769,19 @@ def tp_getattro_impl(
...
+                result = result.clone(source=AttrSource(self.source, name))
```
Comment (aorenste) [inline torch/_dynamo/variables/tensor.py]: [cleanup] My agent says:
> Optional freebie, since this line is already cloning: dropping the inherited location lets the fresh object pick up one from the frame being compiled.
>
> ```python
>                 result = result.clone(
>                     source=AttrSource(self.source, name), source_location=None
>                 )
> ```
>
> `clone()` copies `__dict__`, so as written the new constant inherits the singleton's `source_location` -- which `push()` pinned, process-wide and permanently, to whichever frame first executed `LOAD_CONST None`. Passing `source_location=None` leaves it for `push()` to fill in locally. On a plain `p.requires_grad` read that moves the attribution from an unrelated finished compile onto the frame being traced; on the `trace_autograd_ops` `.grad` path it lands on a `polyfills/__init__.py` frame rather than user code, so it's an improvement in provenance rather than a fix. `source_location` feeds `_format_stack_source_attribution`, which lands in the `dynamo_graph_break_reason` tlparse artifact.
>
> No behavior change either way -- the regression test above passes with and without it -- so skip it if you'd rather keep the diff tight.

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

[cleanup] My agent says:
> The invariant this asserts -- "a constant one is not ours to mutate" -- still isn't established. `symbolic_convert.push()` (`symbolic_convert.py:2171-2189`) writes frame-specific state onto the same three singletons:
>
> ```python
> if val.source_location is None:
>     ...
>     val.set_source_location(SourceLocation(filename=self.f_code.co_filename, ...))
> ```
>
> Every `set_source_location` call site is guarded on `source_location is None`, so first writer wins and the location is pinned, for the life of the process, to whichever frame first executed `LOAD_CONST None`. That field is read by `_format_stack_source_attribution` (`symbolic_convert.py:5266`), and its output is embedded in the `dynamo_graph_break_reason` artifact emitted through `trace_structured` -- so this is what a production user reads out of tlparse, not just a console log.
>
> Two frames, the second graph-breaking with a `None` still on the value stack:
>
> ```python
> def first(x):
>     y = None                          # line 5
>     return x + 1, y
>
> def second(x):
>     return None, print("break"), x    # line 12
> ```
>
> `dynamo_graph_break_reason` for the break in `second`:
>
> ```
> Stack variable source attribution:
>   ConstantVariable(NoneType: None) originated from:
>   File "<stdin>", line 5
>
> User code traceback:
>   File "<stdin>", line 12, in second
> ```
>
> Line 5 is in a different, already-finished compilation; the `None` on the stack came from line 12. Pre-existing and not this PR's bug, and `source_location` has no consumer outside that attribution, so there's no correctness impact -- but the artifact confidently points at unrelated user code, which is worse than saying nothing.
>
> Worth noting for whoever picks this up: just refusing the write on the singletons trades the wrong line for a missing one, since `_format_stack_source_attribution` skips entries whose `source_location is None`. Getting it right means keeping per-use state off the shared instance (side table keyed by stack slot, or handing out a fresh constant when a location needs attaching) rather than adding another don't-mutate-this rule.
